pkg/gardenlet/controller/shoot/shoot_control.go

```diff
@@ -213,6 +213,9 @@ func (c *Controller) deleteShoot(shoot *gardencorev1beta1.Shoot, logger *logrus.
 		utilerrors.ToExecute("Check and sync Shoot with Seed", func() error {
 			return c.checkSeedAndSyncClusterResource(shoot, o)
 		}),
+		utilerrors.ToExecute("Ensure ShootState resource for Shoot", func() error {
```

Comment:
What was the motivation to put this here instead of to the top of the files https://github.com/gardener/gardener/blob/master/pkg/gardenlet/controller/shoot/shoot_control_reconcile.go and https://github.com/gardener/gardener/blob/master/pkg/gardenlet/controller/shoot/shoot_control_delete.go ? I think this rather belongs there.

Reply:
My thoughts were that the ShootState is a pretty important component and adding it here means that we won't have to add additional task dependency for every step that would require it in the flow graph.

Tbh I mostly did it because of `c.checkSeedAndSyncClusterResource()` which I was wondering why it was called here (it just does an additional check) and then the cluster resource is synced one more time at the start of the reconciliation flow graph

Reply:
The `Cluster` resource is special because it contains the `.status.lastOperation` of the `Shoot` which might be set to `Failed`. Extension controllers react on this and don't do anything in this case. The gardenlet does neither, i.e., it does not execute any flow, however, it has to sync the `Cluster` object to the seed so that the extensions get to know the status as well.

The `ShootState` belongs in the flow graph, I guess. Yes, it is important, but not "that important" that we require special handling like for the `Cluster` object.